website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,556 @@
+
+* Use reference-counted pointers and a mutex, i.e. `Arc<Mutex<T>>`
+* Use "actors" and connect them via channels, i.e. spawn multiple tasks with loops that read and write to channels
+
+Both of these options have a runtime overhead:
+Locks can result in contention and sending messages through channels requires copying.
+In addition, multiple tasks running inside a runtime operate in a non-deterministic order which can easily lead to race conditions and in the worst-case, dead-locks.
+It appears that with either of these options, we arrive at a design that feels brittle, is prone to dead-locks and no longer employs zero-cost abstractions, yet avoiding all of these is one of the reasons we wanted to use Rust in the first-place!
+
+In the sans-IO world, these problems don't exist.
+Our protocol code doesn't spawn any tasks and thus, `&mut self` is all we need to mutate state.
+Without tasks or threads, we also don't need synchronisation primitives like `Mutex`.
+Without channels, there is no need to copy data: The state machine can simply directly reference the buffer we passed to the socket.
+
+Last but not least, we've also found that ever since we moved to sans-IO, our code became much easier to understand.
+No more tracking down of: Where is the other end of this channel? What if the channel is closed? Which other code is locking this `Mutex`?
+Instead, it is all just nested state machines and regular function calls.
+
+## The downsides
+
+There are no silver-bullets and sans-IO is no exception to this.
+Whilst writing your own eventloop gives you great control, it can also result in subtle bugs that are initially hard to find.
+
+For example, a bug in the state machine where the value returned from `poll_timeout` is not advanced can lead to a busy-looping behaviour in the eventloop.
+
+Sequential actions require more boilerplate.
+In Rust, `async` functions compile down to state machines, with each `.await` point representing a transition to a different state.
+This makes it easy for developers to write sequential code.
+Without `async`, we need to write our own state machines for expressing the various steps.
+How annoying this will be in practise depends on your problem domain.
+Modelling a request-response protocol is not very difficult as we've seen in the example of a `StunBinding`.
+
+Finally, the sans-IO design is not particularly wide-spread (yet) in the Rust community.
+As a result, there are very few libraries out there that follow it.
+Most of them will either implement blocking or non-blocking IO instead of sans-IO.
+
+## Closing
+
+Writing sans-IO code is unusual at first but really enjoyable once you get the hang of it.
+In part, this is because Rust provides great tools for modelling state machines.
+More so, the fact that sans-IO forces you to handle errors as you would any other input simply feels like the way networking code should be written.
+
+Discuss this post on [Hackernews](TODO) or [Reddit](TODO).
```

Comment:
These need to be filled in after publishing.

Reply:
👍

Reply:
Obligatory reminder about linking HN is we don't ever want to share the links directly to avoid the vote-ring detection they have, which will bury the post.

Reply:
So we can't link directly to the comment thread? That is a bit stupid. Then we don't need to coordinate anything around publishing the post.